### DISCUSSION TITLE
Proposal: forward pipe operators

### DISCUSSION THREAD

**Discussion by orthoxerox** [dotnet/csharplang#74] [General]
# Forward pipe operators

## Summary
[summary]: #summary

Operators for passing values to methods/delegates/constructors/arbitrary expressions.

See also https://github.com/dotnet/roslyn/issues/5445, https://github.com/dotnet/roslyn/issues/3171

## Motivation
[motivation]: #motivation

Promoting the use of static (preferably stateless) functions.

## Detailed design
[design]: #detailed-design

### Grammar

```antlr
primary-no-array-creation-expression
    : ... //existing grammar
    | pipe-expression
    | placeholder-expression
    ;

pipe-expression
    : primary-expression pipe-operator primary-no-array-creation-expression
    ;
    
pipe-operator
    : '|>'
    | '?>'
    ;

placeholder-expression
    : '@'
    ;
    
object-creation-expression
    : 'new' type arguments? object-or-collection-initializer?
    ;

arguments
    : '(' argument-list ')'
    ;
```

#### Binding

1. The lhs is bound as an assignment to a temporary local in the nearest scope to ensure it is evaluated first.
2. The rhs is bound depending on the presence of a placeholder:
2.1. The rhs is visited by a placeholder finder
2.1.1. Pipe expressions do not check their rhs for placeholders
2.1.2. Placeholders report they are a placeholder
2.1.3. Other nodes recur
2.2. If placeholders are present, the rhs is bound as is using PipeBinder
2.2.1. Placeholders are bound as BoundLocals by PipeBinder (regular binder binds them as BoundBadExpressions)
2.3. If placeholders are not present, the rhs is bound as an invocation:
2.3.1. If the rhs is an ObjectCreationExpressionSyntax with arguments, it is bound as a BoundBadExpression (tbd)
2.3.2. If the rhs is an ObjectCreationExpressionSyntax without arguments, it is bound as a BoundObjectCreationExpression with a single argument
2.3.3. If the rhs is a ConditionalAccessExpressionSyntax, it is (TODO: explain the logic in the prototyp)
2.3.4. If the rhs is an AwaitExpressionSyntax, (TODO: think about that)
2.3.5. Otherwise the rhs is bound as an invocation expression with one argument
3. The expression is bound with lhs, rhs and the type of operator (conditional or not).

#### Code Gen

1. Assignment of the lhs to the temporary local is emitted and the value is discarded.
2. Code gen depends of the conditional parameter of the BoundPipeExpression:
2.1. If the BoundPipeExpression is conditional:
2.1.1. If the result is `null` or a `Nullable<T>` with no value, a default value of the bound rhs type is emitted
2.1.2. Otherwise the rhs is emitted
2.2. If the BoundPipeExpression is not conditional, the rhs is emitted

## Drawbacks
[drawbacks]: #drawbacks

I don't really see any.

## Alternatives
[alternatives]: #alternatives

None.

## Unresolved questions
[unresolved]: #unresolved-questions

What is the correct precedence of the pipe operators? In my (invalid) prototype I placed them between ternary and coalescing operators, but I can't remember why.

Are there any other syntax nodes that protect placeholders inside? Lambdas/delegates?

Should we have an operator that chains function calls using `SelectMany`?

## Design meetings

None that I know of.

**Comment by alrz** [comment]
Currently `new SomeType1` is not a valid expression, so you would require to change allowed syntax forms in that context. Also, that syntax won't work in case of currying e.g. `e |> F(1)` would be equivalent to `F(1, e)` (note that no actual currying is happening, just it would be allowed syntactically). I think we should require parentheses for  any number of arguments (including zero) to make that consistent with existing production rules of the language. this is what has been proposed in dotnet/roslyn#5445.

**Comment by orthoxerox** [comment]
@alrz Requiring parentheses would violate the principle of least surprise, IMO:

```c#
SomeType e = default(SomeType);

Action<SomeType> F(int x) => blah;
void F(int x, SomeType st) => bleh;

e |> F(1); //calls the second overload
e |> F(1)(); //calls the first overload
```

Also, currying is usually done from the left, not from the right:

```c#
static Func<T2, TResult> Curry<T1, T2, TResult>(Func<T1, T2, TResult> func, T1 t1) 
    => t2 => func(t1, t2);
```

Inserting a parameter in front of the supplied parameters feels weird, I'd rather support something like https://github.com/dotnet/roslyn/issues/3561 or https://github.com/dotnet/roslyn/issues/8670 or https://github.com/dotnet/roslyn/issues/3171

**Comment by alrz** [comment]
While I disagree that it would be any useful to forward to return value of the RHS in context of C#, I think it would help to provide language grammar you have in mind for the forward pipe operator.

**Comment by DavidArno** [comment]
@orthoxerox,

I'm noticing a problem with proposals created by someone other that the team. In the latter case, there is a proposal spec that is linked to that shows the details, such as syntax. Would it be worth creating a gist for example, that could used for the spec for this proposal (same idea would apply to other community proposals too)?

**Comment by orthoxerox** [comment]
@DavidArno I guess I could fork the repo and push the proposals there.

**Comment by DavidArno** [comment]
Ah good point. Hadn't thought of doing that. Excellent idea as a PR from that is easier too.

**Comment by orthoxerox** [comment]
@alrz Could you add some markdown examples of grammar specs to your proposed issue template? This would help a whole lot.

**Comment by alrz** [comment]
In that proposal you could use a general (or restricted, depending on the precedence) expression as RHS since there is no syntactic requirements. Then allowed forms will be checked semantically. e.g only permit expressions that a value can be forwarded to, namely, *invocation-expression*, *object-creation-expression*, an *await-expression* that contains any of applicable expressions (recursively), etc. 

Note that `expression '|>' expression` is ambiguous as it does not specify any precedence for expressions in which the operator is applied, also it wouldn't allow `new SomeType` as it's not a valid expression.

**Comment by orthoxerox** [comment]
@alrz in my case it's not restricted at all, since the pipe expression replaces all invocation-likes: the rhs provides an invocable. I think only object creation expression is weird, since it doesn't actually provide an invocable and has to be syntactically valid without both argument list and initializers now.

I should check if my approach works with await, though.

**Comment by YaakovDavis** [comment]
It would be nice if piping to arg-positions other than the last will be supported. This can be achieved by a new place-holder syntax, e.g.:

```
void M(int a, int b, int c) {}
2 |> M(1, .., 3);
```

Alternative place-holder suggestion:
```
2 |> M(1, @, 3);
```

This is especially useful given that methods in C# codebases typically aren't designed for piping purposes.

**Comment by orthoxerox** [comment]
@YaakovDavis yes, shorthand lambdas are an option as well, but they have very complicated boundary detection rules. Maybe if the placeholders were restricted to argument lists and not arbitrary expressions... but then they wouldn't be as useful as full-powered shorthand lambdas.

**Comment by YaakovDavis** [comment]
@orthoxerox,
> shorthand lambdas

Are you referring to the proposed lambda syntax from https://github.com/dotnet/roslyn/issues/5445?  
I see my proposal as orthogonal to that. Piped-value place-holders could work with, or without the lambda support. Please correct me if I miss something.

**Comment by orthoxerox** [comment]
@YaakovDavis I think you must've linked the wrong proposal. Anyway, yes, I am referring to that syntax. I wouldn't call you proposal orthogonal to the shorthand lambda, since piping to a shorthand lambda would look the same as piping to an invocation with a placeholder:

```c#
2 |> M(1, .., 3);
2 |> M(1, @, 3);
```

This makes placeholders a strict subset of the shorthand lambda functionality. Return values are commonly used as method receivers, not just regular arguments, so if we allow placeholders to be used there as well, they might be much more useful than I originally thought...

**Comment by YaakovDavis** [comment]
@orthoxerox Seems like you're referring https://github.com/dotnet/roslyn/issues/3561. Am I correct?
I wasn't aware of this proposal till now.

Let's continue the discussion after confirming that we're on the same page :)

**Comment by HaloFour** [comment]
I'm not sure why forward pipe operator makes a language more "functional".  More "fluent", maybe.

In my opinion exposing any of these four operators opens a can of worms as you'll soon run into paradigmatic differences between C# and functional languages.  Currying and overloading are two big ones.  Those two features generally can't coexist in the same language (or at least in the same context, which is the line that F# walks).  In F#, piping an expression into a function that accepts multiple arguments results in a currying operation.  That clearly can't happen here.  So you still end up with having arguments having to be defined and embedded at the call site.

Personally I prefer extension methods.  They achieve roughly the same functionality but fit more into the C# syntax.  Hopefully local extension methods will be supported at some point.

**Comment by alrz** [comment]
@HaloFour In my opinion, this should not be perceived as a move to make the language more "functional". That's just a construct to write more readable call chains with the same lexical order as they will be executed or without leaving the expression context.

**Comment by gafter** [comment]
@orthoxerox 
> If the rhs is a lambda declaration, it is optimized away by inlining the body.

That doesn't make any sense, as the lambda would be an error by virtue of not having a (delegate) target type. (You can't "optimize" an error)

Does anyone have a plausible concrete use case for this that demonstrates that it would be of value in frequently occurring code patterns?

**Comment by HaloFour** [comment]
@gafter 

> You can't "optimize" an error

I beg to differ.  Isn't that called "fail fast"?

**Comment by orthoxerox** [comment]
@gafter I guess this snippet is a stub from an older version of the proposal where shorthand lambdas were included. The lambda expression would be rewritten before being bound, so it wouldn't be an error yet. With longhand lambdas this doesn't make much sense, I agree.

**Comment by orthoxerox** [comment]
@HaloFour the presence of `|>` nudges the developers towards writing pure static functions because they are easy to chain.

I'm kinda warming up to the idea of adding placeholders to the proposal, to offset the impossibility of currying.

**Comment by MgSam** [comment]
I think placeholders to indicate where to use the value being piped are necessary. Eliding or shortening the argument list of a method looks odd and would be no doubt be confusing and introduce potential ambiguities.

While its certainly possible to create similar behavior to this in today's C# using an extension method that accepts a `Func<T1, T2>`, it is, of course, not nearly as efficient as the proposed pipe operator could be as it requires closing over any local variables you want to use.

@gafter Any nested function calls seem a pretty great use case for this. Reading function calls from the inside out makes it unnecessarily hard to follow complex logic.

``` C#
var foo = Substring(ParseJson(LoadData())[5].Value, 0, 3)
```
vs
``` C#
var foo = LoadData() |> ParseJson(@)[5] |> Substring(@, 0, 5)
```

In fact, I wish the MS Excel guys would take a page from F# and add this operator already.  Trying to decipher tons of nested function calls is a nightmare.

**Comment by orthoxerox** [comment]
Updated the OP with grammar and more detailed design, including placeholders.

**Comment by alrz** [comment]
@gafter AFAIK this was championed by dotnet/roslyn#5445's assignee as it's labeled "Ready", no? If so, is there a spec draft somewhere that we can base the discussion on it to ensure that we're on the same page?

**Comment by AlphaDriveLegal** [comment]
Hi All,
I was wondering why the |> operator was picked for this.  My thought was that something like -> might be easier to understand.  I didn't see anyone else mentioning that here so I just wanted to throw it out:

````
var results = 
  from x in items
  where x != null && x.Age > 21
  select x
  ->Skip(5)->Take(10)->ToList()
  ;

````

**Comment by orthoxerox** [comment]
@AlphaDriveLegal https://github.com/dotnet/roslyn/issues/5445#issuecomment-143351517 and https://github.com/dotnet/roslyn/issues/5445#issuecomment-151205058

**Comment by gafter** [comment]
@alrz That proposal is different, as it has the input parameter added to the argument list. This proposal (in its first iteration, at least) requires there be no argument list.

**Comment by alrz** [comment]
@gafter Should I open another issue for that one or this suffices?

**Comment by gafter** [comment]
@alrz You can hold discussion wherever you want. If you have a different proposal to discuss then it makes more sense to do so on a different issue.

**Comment by orthoxerox** [comment]
@alrz Why do you think that argument lists are better than no argument lists?

**Comment by gulshan** [comment]
I support the idea of placeholders. I prefer to reuse `_` here. The **gap to be filled** with supplied data feels more explicit with `_` than `@` to me-

```csharp
var foo = LoadData() |> ParseJson(_)[5] |> Substring(_, 0, 5);
```

**Comment by YaakovDavis** [comment]
> I prefer to reuse _ here.

`_` is a valid identifier, and hence problematic as a place holder.

**Comment by orthoxerox** [comment]
@gulshan this will conflict with discards:

`getFirst() |> foo is Foo(_, 2)`

**Comment by YaakovDavis** [comment]
I implemented in a personal language the following piping syntax, which might be a good fit for C# as well:

```
var foo = LoadData() .> ParseJson(.)[5] .> Substring(., 0, 5);
```

`.>` is easier to type on the keyboard than `|>` (try it now), and `.` as a place-holder resonates well with it.

As we're at it, I also support the following:

```
// The following is equivalent to a .> c(b) .> d(), which is equivalent to d(c(b, a)).
// Notice the double-dot. Arbitrary "levels" can be used.
a ..> b .> c() ..> d() 
```

This allows minimizing nested-invocations, though arguably can be less readable in some scenarios.

**Comment by gafter** [comment]
@orthoxerox I didn't express a preference. Perhaps if there were some realistic use cases for either one or the other I'd have a reason to prefer something.

**Comment by alrz** [comment]
> Why do you think that argument lists are better than no argument lists?

@orthoxerox Mainly because C# does not have currying and removing argument lists would limit potential use cases and lessen the usefulness of the feature overall.

**Comment by orthoxerox** [comment]
@alrz I've added placeholders to the existing proposal. They provide that flexibility C# needs and do it without ambiguity. Since putting `(@)` everywhere would only generate syntactic noise, argument lists containing only a single placeholder can be omitted.

**Comment by alrz** [comment]
Honestly I dislike that idea. In my opinion, placeholders wouldn't fit the purpose of value forwarding. I think that would make more sense in the context of delegate generators (dotnet/roslyn#3561).

**Comment by orthoxerox** [comment]
Shorthand lambdas are something I like too, but I feel that they will meet even more resistance from the LDT than pipes, since that will be the third way to write an anonymous delegate. They are a superset of placeholders so they can be added later, though.

By the way, where does a shorthand lambda begin or end? Longhand lambdas begin after `=>` and are one expression long. Swift wraps them in `{}`. I know Scala's shorthand lambdas have no visible boundaries, but I have no idea how it does that (Idris finds out the boundaries of a lifted expression the same way, I presume).

**Comment by DavidArno** [comment]
To add yet another idea to the mix here, the approach I tried to take myself, using `Pipe` extension methods, rather than eg the `|>` symbol was to use tuples, rather than worrying about placeholders, parameter order, etc. I couldn't implement it, due to [the Roslyn bug, 16756](https://github.com/dotnet/roslyn/issues/16756), but the general idea (using some previous examples) would be:
````cs
(b, a) |> c |> d; // Equivalent to d(c(b, a)).
var foo = ((LoadData(), 5) |> ParseJson, 0, 5) |> Substring;
````

So as a general rule, `f(p1 ... pN)` becomes `(p1 .. pN) |> f`

**Comment by orthoxerox** [comment]
@DavidArno I am not sure if the nested tuples make it more readable than the original method nesting. I can put meaningful line breaks in pipes with placeholders and not very meaningful line breaks in regular chains with mixed fluent and nested calls, but I can't find a way to format pipes with nested tuples in an acceptable way that makes it easy to add or remove steps or at least understand what's going on.

```c#
var foo = ((LoadData(), 5) |> ParseJson, 0, 5) |> Substring; //where should I insert the line breaks?

var foo = LoadData()
    |> ParseJson(@, 5) 
    |> @.Substring(0, 5);

var foo = ParseJson(LoadData(), 5)
    .Substring(0, 5);
```

**Comment by YaakovDavis** [comment]
> way that makes it easy to add or remove steps

It's also easy to place breakpoints on specific steps in the pipeline with the piping approach, whereas it's difficult with nesting.

**Comment by DavidArno** [comment]
@orthoxerox,
````cs
var foo = ((LoadData(), 5) 
    |> ParseJson, 0, 5)
    |> Substring;
````
Yep, I'd agree that readability does suffer. But it's all part of my secret plan to turn C# into Lisp... 😆

**Comment by orthoxerox** [comment]
@DavidArno you should just switch to https://github.com/jantolenaar/kiezellisp then. :grin:

**Comment by gulshan** [comment]
I want to add a few things-

- When there is only one argument `(@)`, only method name should be enough without parenthesis.
- `await` within the chain.
- `using` within the chain. The disposal will happen after the evaluation of the expression. Have to be careful here.
- `return` within the chain.
- Ability to use `our var` int the subsequent chain calls.

With these features this code of reading an int from a file -

```csharp
var buffer = new Byte[32];
using (FileStream fileStream = File.Open(fileName, FileMode.Open))
{
    await fileStream.ReadAsync(buffer, 0, buffer.Length);
    var str = Encoding.Unicode.GetString(buffer);
    var success = Int32.TryParse(str, out int num);
    return success ? num : -1;
}
```

can become this -

```csharp
var buffer = new Byte[32];

fileName
|> using File.Open(@, FileMode.Open)
|> @.ReadAsync(buffer, 0, buffer.Length)
|> await;

buffer
|> Encoding.Unicode.GetString
|> Int32.TryParse(@, out int num)
|> @ ? num : -1
|> return;
```

**Comment by gulshan** [comment]
Another example with `using` within the chain. At first this extension method `Apply` is needed-

```csharp
public static T Apply<T>(this T obj, Action<T> action) { action(obj); return obj; }
```

Then this code with heavily nested `using` blocks of #114

```csharp
public void Foo(String connectionString) {
    using (var connection = new SqlConnection(connectionString)) {
        connection.Open();
        using (var command = connection.CreateCommand()) {
            command.CommandText = "SELECT FOO FROM BAR";
            using (var reader = command.ExecuteReader()) {
                while (reader.Read()) {
                    ProcessRecord(reader);
                }
            }
        }
    }
}
```

can become-

```csharp
public void Foo(String connectionString)
{
    connectionString
    |> using new SqlConnection(@)
    |> @.Apply(connection => connection.Open())
    |> using @.CreateCommand()
    |> @.Apply(command => { command.CommandText = "SELECT FOO FROM BAR"; })
    |> using @.ExecuteReader()
    |> @.Apply(reader => {
        while (reader.Read())
        {
            ProcessRecord(reader);
        }
    }
}
```

It seems there are still some readability concerns though.

**Comment by mattwar** [comment]
What happens if I have:   ``` f(a(x), b(y))``` ?

**Comment by YaakovDavis** [comment]
@gulshan 
> `|> using new SqlConnection(@)`

It's unclear from the snippet above what's the scope of the `using` statement. 

As `usings`s are specifically about defining scopes, I'm not sure it makes sense to allow piping them.

**Comment by orthoxerox** [comment]
@mattwar 

`x |> a |> f(@, b(y))` or `y |> b |> f(a(x), @)` or even `f(a(x), b(y))`, it depends on which one expresses the underlying logic better for given `f`, `a`, `b`, `x` and `y`.

**Comment by gulshan** [comment]
@YaakovDavis As I have said, the scope of `using` will be the chain. The disposal will happen after the evaluation of the current full forwarding expression. So,

```csharp
using File.Open(fileName, FileMode.Open) |> @.ReadAsync(buffer, 0, buffer.Length) |> await;
```

scope of using ends at the end of line here. And at the last SQL reader example, all three using scope will end at the end of forwarding expression. That's why the loop over reader with while could not be separated from the forwarding expression.

@mattwar @orthoxerox  I think `f(a(x), b(y))` can possibly become `(x, y) |> (a, b) |> f`.

**Comment by orthoxerox** [comment]
@gulshan I don't think that it should be necessary to be able to rewrite every single invocation as piping. Piping is good for expressing long chains of data transformation, with data passed as one parameter. Why invent ways to use piping in other cases if they are not making the intent clear, is it just to avoid naming locals?

**Comment by gulshan** [comment]
I was exploring how far can we go. 😃

**Comment by mattwar** [comment]
@orthoxerox I was subtly point out that piping only feels like a syntactic improvement in situations where you have a series of single argument functions:  ```d(c(b(a)))``` which can be translated to ```a |> b |> c |> d``` because when you introduce multiple argument functions in expressions with additional nested expressions for each argument, and so on, the piping operation only seems to add to the confusion:  ``` k(c(a, b), g(e, f), h(i,j))```  becomes  ```a |> c(@, b) |> k(@, g(e, f), h(i, j))```.  This doesn't read better.  It can get much worse:  ```a |> c(@, b) |> k(@, e |> g(@, f), i |> h(@, j))```

**Comment by orthoxerox** [comment]
@mattwar I won't disagree with that. However, how often do you write something like `k(c(a, b), g(e, f), h(i,j))`?

**Comment by mattwar** [comment]
@orthoxerox not so much, but a lot more often than I write code that could be made clearer with a pipe operator.

**Comment by HaloFour** [comment]
@mattwar 

I have similar concerns.  I don't recall an overabundance of scenarios in my code where I write anything like `d(c(b(a)))`.  Much more frequently there are more parameters involved, often calculated in their own right.  And in those cases I intentionally write out their values to local temporary variables because frankly I'd like to be able to read my own code again in the future and don't subscribe to the notion that compiled code runs faster the fewer newlines are in the original source.

I would further wager that in order to produce pipeline-friendly APIs would involve writing helper functions to existing API.  I already do this with extension methods, but I don't have to argue over the finer points of how additional parameters are supposed to work because I can explicitly design my extension methods to account for this.

I don't really see how well this operator fits into a language like C# that lacks currying (and would likely never support currying due to overloading) and without the other piping and compositional operators from that family.

**Comment by YaakovDavis** [comment]
@HaloFour 
> Much more frequently there are more parameters involved, often calculated in their own right. 

If I may interject,
This is exactly where piping with place-holders helps. You still get to treat those temporary variables individually, you just don't need to name them. 

I see currying as a restrictive, incomplete implementation of piping. Piping IMO is about communicating evaluation order & dependencies *within the codebase*, not about runtime behavior. 

Frankly, "runtime piping" (ie. currying) complicates codebases, and hurts performance of both compilation & runtime, without gaining much in expressiveness. You often have to write un-natural APIs, to satisfy the "last-parameter" notion.
In languages like F#, you may run into bugs where you hold a function, instead of a final value, because you forgot to provide all the parameters. The compiler isn't able to assist, because it has no idea if you meant to curry or to eval. Most of the time you *do* want to eval.
With compile-time piping this can't happen.

**Comment by gulshan** [comment]
@orthoxerox I think some decent examples are needed to show the use-cases of this proposal. The ones I put were quite extreme.

**Comment by alrz** [comment]
Looking a little closer, the proposed `@` syntax is exactly what #91 is trying to accomplish.

Note: I'll use [Mathematica](https://reference.wolfram.com/language/tutorial/PureFunctions.html) syntax for #91; using `#` for parameters and `&` to terminate.

For example,
```cs
.Select(# + 1 &)
// equivalent to
.Select(x => x + 1)
```
So, using the same syntax in the context of forwarding would look like this (#96 syntax for RHS):
```cs
value 
|> $"the value is {#}" &
|> Console.WriteLine(); 
```
Delegates and method references (#84) could also be used (which looks like what's proposed here for RHS),
```cs
value
|> Console.WriteLine; 
```

I think this closes the gap between #96 and #74.

/cc @orthoxerox @HaloFour

**Comment by gulshan** [comment]
I imagine `|>` being a postfix operator for function/method calling. Meaning `"abc" |> Write` is just `Write("abc")`. Thus I proposed to remove the parenthesis after the method-name. That will also help to reduce ambiguity with overloading. And my proposal of using `await` and `return` with `|>` comes from the notion of it being postfix. Using `using` within the chain was a bit too far of an idea.

**Comment by orthoxerox** [comment]
@gulshan the parens were the reason why @alrz opened his own proposal.

@alrz yes, originally two ideas were combined in my roslyn proposal. Without a delimiter shorthand lambdas were a pain to parse, so I dropped them. I'm not sure if a *terminator* is the right option there, though. I think a *prefix* would work better with the existing parser.

**Comment by orthoxerox** [comment]
I've implemented a prototype that combines pipes with placeholders and shorthand lambdas. No conditional pipe support yet, though. Will push it tomorrow when I get home.

**Comment by MgSam** [comment]
Interesting reading- ECMAScript is considering a similar proposal: https://github.com/tc39/proposal-pipeline-operator.

**Comment by mharasimowicz** [comment]
I support this proposal, it can simplify code when flow is clear.

BTW: It's possible to write similar functionality today by writing own delegate classes and overloading | operator, but code is quite verbose, which in practice defeats it's purpose.

I've created gist with example code.

https://gist.github.com/mharasimowicz/a4ff68ab4b890d8b866dcdbd6795cc3b

**Comment by gulshan** [comment]
The ECMAScript proposal made `async` part of the pipeline. Feels great!

**Comment by daiplusplus** [comment]
One thing I'd love to use this for is for "postfix-casts". It feels unnatural to me that the C# cast `(T)` operator  is prefix, not postfix, because it disregards the expected left-to-right evaluation order for non-parenthesised operations, which is how Linq's `Cast<T>()` method works.

Compare:

    ((Int32)enumerable.Select( foo => foo.Bar ).SingleOrDefault() ?? 123).ToString();

With:

    enumerable.Select( foo => foo.Bar ).SingleOrDefault() ?? 123 |> (Int32)@ |> @.ToString();

I feel `|> (Int32)@ |>` syntax could be improved somewhat - any thoughts?

**Comment by dzmitry-lahoda** [comment]
I do fun programming again with `.Do` & `?.Do`.

```
        public static T Do<T> (this T target, Action<T> action)
        {
            action(target);
            return target;
        }

        public static TResult Do<T, TResult> (this T target, Func<T, TResult> func) => 
            func(target);
```

**Comment by dadhi** [comment]
@dzmitry-lahoda,

Agreed.

For me the main benefit is the postfix notation which naturally fits to my train of thoughts / transformation chain from input to required result. This chain naturally flows forward - it is where the piping or your extension methods fit.

Given the usual notation I often need to return back to store the result of first transformation, then jump forward again to use the result.

It may not need to be applied everywhere, but it helps me a lot in small methods or LINQ queries, to produce a cleaner code. Usually it enables to convert such methods from statement to exptession-body, cutting more noise.

Even if it is only may be applied to single parameter methods, that's fine. One improvement at time!

```

R Blah(T input)
{
    var key = GetKey(input);
    var value = FindValue(key);
    return value == null ? null : Process(value); // null propagation cannot be applied here
}

```

Extension methods:

```

R Blah(T input) =>
    GetKey(input).Do(FindValue)?.Do(Process);

```

Pipe operator(s):

```

R Blah(T input) => 
    input |> GetKey |> FindValue ?> Process;

```

**Comment by masaeedu** [comment]
Instead of tying placeholders and the pipeline operator together (at cost of increased complexity to both), can we instead have two simpler operators?

```c#
v |> f |> g |> h
// => always deugars to (and typechecks/dispatches identically to)
h(g(f(v)))
// without regard to number of arguments or overloads of f, g, h
```

In order to deal with multi-argument functions you can explicitly do:

```c#
v |> f |> (x => g(42, "foo", x)) |> ...
```

If the simple lambda wrapping `x => ...x...` feels too tedious, we can have an independent language feature for making partial application more convenient:

```c#
f(?, "foo")
// => desugars to
x => f(x, "foo")
```

Now you can use the two features in isolation or together:

```c#
v |> f |> g(42, "foo", ?) |> ...
```

Of course special casing the emit of `g(42, "foo", ?)` to skip creation of a lambda when inside a `|>` pipeline (for performance reasons) is fine, so long as the runtime semantics are identical.

**Comment by theunrepentantgeek** [comment]
The problem with resorting to lambdas is that they always allocate - and they often capture.

This is going to create a lot of heap debris:
```
v 
|> f 
|> x => g(42, "foo", x) 
|>
```

Whereas, with the original proposal, this would not:
```
v 
|> f 
|> g(42, "foo") 
|>
```

**Comment by masaeedu** [comment]
I guess it's too idealistic to hope there's some emit or JIT optimization that makes `(x => g(42, "foo", x))(v)` cost the same as `g(42, "foo", v)`?

**Comment by theunrepentantgeek** [comment]
AFAIK, there isn't at present - and I believe designing a language feature in a way that it requires CLR changes dramatically changes the likelihood of implementation.

**Comment by orthoxerox** [comment]
@masaeedu @theunrepentantgeek that doesn't have to be a CLR change. `(x => g(42, "foo", x))(v)` is an invalid expression in C# and it can be lowered to a sequence expression `(var x = v; g(42, "foo", x))` by the compiler.

**Comment by masaeedu** [comment]
@HaloFour 

> I don't recall an overabundance of scenarios in my code where I write anything like d(c(b(a)))

If you've written code like `foo.Select(x => x + 1).Where(x => x < 42).SelectMany(Factors).Distinct()`, you've written code like `d(c(b(a)))`.

**Comment by HaloFour** [comment]
@masaeedu 

> If you've written code like `foo.Select(x => x + 1).Where(x => x < 42).SelectMany(Factors).Distinct()`, you've written code like `d(c(b(a)))`.

No, that's writing code like `a().b().c().d()`.  The fact that it's converted into nested static method calls by the compiler doesn't really matter.  The same would be true of any pipe operator.

**Comment by masaeedu** [comment]
The `Select`, `Where`, `SelectMany` etc. are static methods, the fact that they're sugared for `.Whatever()` application is what doesn't really matter here. There are many more such functions that I'd like to be able to apply to values in a chain which can't be given fixed definitions as static methods, but are instead more conveniently computed from other functions and values.

**Comment by daiplusplus** [comment]
@theunrepentantgeek 

> The problem with resorting to lambdas is that they always allocate - and they often capture.

Is allocation on every invocation, or is a single capture-container object created on-demand and shared by subsequent calls (a-la `switch`'s `Dictionary`)?

...and wouldn't the JIT inline the method invocations anyway?

**Comment by theunrepentantgeek** [comment]
I believe that it depends on whether the lambda involves any captures. If it doesn't capture, I believe it will be allocated once. But if it does capture, I'm pretty sure that lambdas are separately allocated for every execution.

**Comment by TahirAhmadov** [comment]
I would warn against writing code in this manner. Often, you want to be able to see the intermediate results of the methods. Debugging code line by line is made difficult by these complex constructs. 
Yes I know assigning local temp vars potentially creates issues with re-assignability, but to solve that, let's continue pushing to add readonly locals.
In optimized release build, this can be optimized to avoid creation and assigning of locals anyway, so this doesn't really improve performance, either.

**Reply by 333fred** [reply]
This problem is solvable, as F# has already proven, and as we've already done with switch expressions in C#. I wouldn't use that as an argument against the feature.

**Reply by CyrusNajmabadi** [reply]
> Often, you want to be able to see the intermediate results of the methods. Debugging code line by line is made difficult by these complex constructs.

I don't see why that would be the case.  It's not like the debugger cares one way or the other :)

**Reply by 333fred** [reply]
> I don't see why that would be the case.

Fluent APIs are this way today, because we don't emit sequence points between the methods. We could do so, but it would have other negative consequences on debug codegen. However, for the pipe operator, sequence points should be fine.

**Reply by TahirAhmadov** [reply]
Does that mean something like step-over command (F10 in VS) would go from method to method in such a scenario, and not jump over the whole thing?
Still, though, even if the return of the previous method is shown in Locals or Auto windows, the returns of previous methods are not.
And I'm not sure what switch expressions have to do with this at all.

**Reply by CyrusNajmabadi** [reply]
> However, for the pipe operator, sequence points should be fine.

Right.  this is a sequence of steps just written in a certain fashion.  Sequence points should work fine.

**Reply by CyrusNajmabadi** [reply]
> Does that mean something like step-over command (F10 in VS) 

The debugger is open for flexibility here.  It can step over, step into, step-to-specific, etc.  But that's a tooling issue we can put effort into regardless of teh lang choices here.

> And I'm not sure what switch expressions have to do with this at all.

They're also an expression that have many interior expressions you might want to hit.

**Reply by TahirAhmadov** [reply]
> They're also an expression that have many interior expressions you might want to hit.

Yes but the semantics are different; one could put some "heavy" methods into switch expressions, but that goes against the "style" of an expression. And with simple expressions, there is no need to step through a switch expressions.
Whereas with this pipe proposal, the intent from the get go is calling "heavy" methods; which is why debugging is much more of a need than with switch expressions.
At the very least, VS should get auto-fixers which convert from this syntax to intermediate locals approach and back and to have a code style preference. That way, even if I start working on somebody else's code and experience difficulties debugging it, I can quickly "unroll" this into the classic syntax.

**Reply by CyrusNajmabadi** [reply]
> Yes but the semantics are different; one could put some "heavy" methods into switch expressions, but that goes against the "style" of an expression.

I don't believe that to be the case at all.  In roslyn we certainly have no such 'style' forour switch expressions.

**Reply by CyrusNajmabadi** [reply]
> That way, even if I start working on somebody else's code and experience difficulties debugging it, I can quickly "unroll" this into the classic syntax.

i feel like you're working backwards here :)  you've presumed a certain thing about tooling, and then you want to backport that into the lang.  Not only do i not agree with teh tooling judgements, i def don't agree on using those judgements in the design as you have here :)

**Reply by TahirAhmadov** [reply]
Well with switch expressions, now that you brought them up, it's already impossible to edit and continue methods containing those. And I suspect that limitation is caused by the way the language compiles them. From where I see it, concerns about negative impacts on tooling caused by language choices are justified - in some cases. (If not in this case - great - but I figured I need to ask anyway.)

**Reply by CyrusNajmabadi** [reply]
> And I suspect that limitation is caused by the way the language compiles them.

It isn't.  It's due to us not having added support to EnC for them :)

> From where I see it, concerns about negative impacts on tooling caused by language choices are justified

Only in the sense that each feature we do has a tooling cost.  But by this measure, then we could not do any features because they're always going to take work to get through the entire tooling chain. :)

**Reply by CyrusNajmabadi** [reply]
Note: we'd also happily take community contributions to help out here :)

**Reply by 333fred** [reply]
> From where I see it, concerns about negative impacts on tooling caused by language choices are justified - in some cases

You're not wrong (see editing source generators). However, I don't believe this to be one of those cases. Switch expression patterns cannot be edited because they are compiled into a DAG and then topologically sorted: there's no way to relate the input and the output. However, the expressions here are not transformed in a complex manner, and enc support, while not _trivial_, should not be a problem.

**Reply by Xyncgas** [reply]
I sensed a C++ 'move' semantic here

**Comment by mdsmithson** [comment]
I’m adding to this discussion to advocate for the forward pipe operator (|>) in C#, emphasizing how its lambda calculus-inspired design is an honest, innovative step for the language’s evolution. By prioritizing functional-style data flows (e.g., data |> Transform1 |> Transform2), |> enhances C#’s expressiveness and creativity, aligning with its multi-paradigm future. Non-critical disputes about articulation or performance are less pertinent, as these are relative and can be optimized, while the paradigm shift offers lasting value.

Why Lambda Calculus-Inspired Design Matters

Rooted in Alonzo Church’s lambda calculus (1930s), which formalized computation through functions, |> brings C# closer to the elegance of stateless, composable pipelines. Church’s work, alongside his student Alan Turing’s Turing machine, birthed modern computing via the Church-Turing thesis. Lambda calculus influenced functional languages like F#, where |> thrives, and its adoption in C# signals a commitment to innovative, readable code. Unlike inheritance-heavy designs, |> fosters linear, predictable data transformations, reducing complexity and aligning with modern programming trends (e.g., reactive pipelines).

While Turing’s Enigma-breaking work at Bletchley Park leveraged computational principles, the broader legacy of lambda calculus—enabling powerful, abstract systems—has a dual edge. It fathered advancements but also dystopian risks like surveillance and AI sprawl. In C#, |> counters this by promoting clarity and simplicity, making complex systems more manageable.
